### PR TITLE
Feature/160664041 verify jwt

### PR DESCRIPTION
**Pull request by meaning-matters**
Please check this work.
I'll add more tests plus the expiry time check this weekend (this is easy and can be copied from Ugo's work on Android).

### CODE REVIEW

File: JWT/JWTTools.swift [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    func now() -> Date
+}
+
+/// Class with JWT related functionality.
+public struct JWTTools
+{
+    private struct Constants
+    {
+        static let signatureSize = 64
+        static let timeSkew = TimeInterval(0.3)
```
Comment (mirceanis) [inline JWT/JWTTools.swift]: this skew should be 5 minutes.

Comment (meaning-matters) [reply JWT/JWTTools.swift]: I misread the Android code and already found 300ms extremely short. Glad I was 'wrong'.

File: Example/Tests/JWTToolsTests.swift [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            XCTAssertTrue(false)
+        }
+    }
+
+    func testFourPartsException()
+    {
+        do
+        {
+            _ = try JWTTools.decode(jwt: "header.payload.some.more")
```
[the remark this answers is not in the document]
Comment (meaning-matters) [reply Example/Tests/JWTToolsTests.swift]: Will write more tests, covering all other possible errors later.

File: JWT/JWTTools.swift [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    private static var universalResolver: UniversalDIDResolver?
+    {
+        var resolver = UniversalDIDResolver()
+
+        let ethrResolver = EthrDIDResolver(rpc: JsonRPC(rpcURL: Networks.shared.rinkeby.rpcUrl))
```
[the remark this answers is not in the document]
Comment (meaning-matters) [reply JWT/JWTTools.swift]: Added a TODO.

File: JWT/JWTTools.swift [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            throw JWTToolsError.malformedNotBase64
+        }
+
+        guard signature.count == Constants.signatureSize ||
+              signature.count == Constants.signatureSize + 1 else
+        {
+            throw JWTToolsError.invalidSignatureSize(signature.count)
+        }
+
+        try JWTTools.checkDates(payload: payload)
```
[the remark this answers is not in the document]
Comment (meaning-matters) [reply JWT/JWTTools.swift]: Yeah good point.

File: JWT/JWTTools.swift [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+        guard !parts[2].isEmpty else
+        {
+            throw JWTToolsError.malformedEmptySignature
+        }
+    }
+
+    private static func checkDates(payload: JWTPayload) throws
+    {
+        if let issueDate = payload.iat, issueDate > (now() + Constants.timeSkew)
```
[the remark this answers is not in the document]
Comment (meaning-matters) [reply JWT/JWTTools.swift]: A `Date` + `TimeInterval` results in a `Date`: https://developer.apple.com/documentation/foundation/date/2293740

File: JWT/JWTTools.swift [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        let signatureBase64 = try JWTTools.base64urlToBase64(base64url: parts[2])
+
+        let header = try JWTHeader(dictionary: JWTTools.base64ToDictionary(base64: headerBase64))
+        let payload = try JWTPayload(dictionary: JWTTools.base64ToDictionary(base64: payloadBase64))
+        guard let signature = Data(base64Encoded: signatureBase64) else
+        {
+            throw JWTToolsError.malformedNotBase64
+        }
+
+        guard signature.count == Constants.signatureSize ||
```
[the remark this answers is not in the document]
Comment (meaning-matters) [reply JWT/JWTTools.swift]: Done

**Pull request merged**